### PR TITLE
Updating downloader subpackage

### LINKED ISSUE #99: selecting maps from corpus for testing or annotation
**Is your feature request related to a problem? Please describe.**
Right now, MapReader assumes you want to download map images from a tile server service based on knowledge of the coordinates contained within a map. 

In more detail:
download tileserver function (not it's proper name...) - option for downloading maps from queries (based on coordinates) or all. There are also `index` numbers which makes it seem like you can select from within an index, but this didn't seem to work (@rwood-97 tested).

However, often users want to make a selection from within a large set of maps rather than downloading a whole tile layer. There are couple of use cases for this:

- when testing MapReader for the first time, working with a limited dataset is handy
- when selecting images for the `annotation_set` (right now this is not documented, assumes user will do this manually)

**Describe the solution you'd like**
Ability to say "I would like to select these maps with metadata A, B, C, D from tile layer for testing or for annotation_set."

**Describe alternatives you've considered**
Two alternative ideas

Easy workaround: use geojson.io to view maps and select sheets.

Another idea: Use Observable notebook with tile layer + metadata file to interactively select maps based on certain metadata/visual features.

**Additional context**
None at moment.

### LINKED ISSUE #108: Download map by number(s)
**Is your feature request related to a problem? Please describe.**

My feature request is not related to a problem, only an extension on the additional code.

**Describe the solution you'd like**

When loading the data, the comment `my_ts.plot_metadata_on_map(add_text=True)` adds a number to each map. See image below. Would it be possible to query by number id instead of lat,long? Or if this already exists, maybe add this to the documentation?

image

**Describe alternatives you've considered**

None.

### PR DESCRIPTION
**Pull request by rwood-97**
### Summary

Updates to the downloader subpackage to allow a variety of different methods for downloading map sheets.
Fixes #108 
Fixes #99 

### Describe your changes
Two classes when downloading maps now:
1) SheetDownloader - for if you have metadata and want to download and merge tiles from a tileserver which correspond to each map sheet
2) Downloader - for if you don't have metadata but want to download and merge tiles from a tileserver using a polygon (area) to define each 'map image'

SheetDownloader contains methods for downloading :
- All map sheets in metadata
- Map sheets overlapping with a defined polygon (area)
- Map sheets completely within a defined polygon (area)
- By WFS index no. (see pic below)
- Map sheets that contain a single point (coordinate)
- Using query results list (queries can be any of the above methods but are essentially done without downloading)

Download code now taken from https://github.com/baurls/TileStitcher.
(Maybe can bring in some of old code to improve download speed but will need to look into this a bit more - RW).

### Checklist before assigning a reviewer (update as needed)
  
- [x] Self-review code
- [x] Ensure submission passes current tests
- [x] Add tests
  
### Reviewer checklist
  
Please add anything you want reviewers to specifically focus/comment on.

- [ ] Everything looks ok?

### CODE REVIEW

**Comment by rwood-97** [comment]
Need to add download sucess options:
- [] Only append to metadata.csv if download was successful
- [] What if file already exists?

**Comment by rwood-97** [comment]
- [ ] add option for querying/downloading any map sheet whose title [or other metadata field to be defined by user] contains X word(s)

**Comment by rwood-97** [comment]
- [ ] Add download_utils to init

**Comment by rwood-97** [comment]
Do we think example this seems correct? 
Generally left and right are same size but top and bottom are slightly varied - why?

> [INFO] Size in meters of left/bottom/right/top: 29715.82/37687.15/29715.82/37953.04
> Each pixel is ~5.047 X 5.094 meters (height x width).
> [INFO] Size in meters of left/bottom/right/top: 31119.71/39470.62/31119.71/39743.08
> Each pixel is ~5.285 X 5.335 meters (height x width).
> [INFO] Size in meters of left/bottom/right/top: 28105.17/39223.74/28105.17/39470.62
> Each pixel is ~5.228 X 5.300 meters (height x width).
> [INFO] Size in meters of left/bottom/right/top: 30722.03/38965.38/30722.03/39236.07
> Each pixel is ~5.218 X 5.267 meters (height x width).
> [INFO] Size in meters of left/bottom/right/top: 28742.65/38195.66/28742.65/38451.32
> Each pixel is ~5.103 X 5.162 meters (height x width).

File: mapreader/download/sheet_downloader.py [hunk 1]
```diff
@@ -0,0 +1,1097 @@
...
+        ----------
+        zoom_level : int, optional
+            The zoom level to use when creating the grid bounding box.
+            Later used when downloading maps, by default 14.
+        """
+        if not self.polygons:
+            self.get_polygons()
+
+        if self.crs != "EPSG:4326":
+            raise NotImplementedError("[ERROR] At the moment, MapReader can only create grid bounding boxes and download map sheets projected in EPSG 4326.")
```
Comment (andrewphilipsmith) [inline mapreader/download/sheet_downloader.py:110]: I'm very happy that this is an explicit error message. However, for accuracy/pedantry, the error message should be:

"[ERROR] At the moment, MapReader can only create grid bounding boxes and query map sheets, using coordinates in WGS1984 (aka EPSG:4326)."

**Review by andrewphilipsmith — COMMENTED** [review]
Great work.

The only real thing is the repeated hard-coded temp directory name for tile downloads. Otherwise, merge when you're ready.

File: tests/ignore_test_non_geo_pipeline.py
[the remark this answers is not in the document]
Comment (rwood-97) [reply tests/ignore_test_non_geo_pipeline.py:1]: Yep, I made this a separate issue to redo the 'pipeline' tests so its still 'ignore_xx' for now (See #171).

File: mapreader/download/sheet_downloader.py (hunk 1 above)
Comment (rwood-97) [reply mapreader/download/sheet_downloader.py:110]: Have done this but also made an extra ticket for us to (at some point) make it possible to use other CRSs by converting to EPSG:4326 in the code.

File: mapreader/download/tile_loading.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            Download urls for tileservers.
+        img_format : Union[str, None], optional
+            Image format used when saving tiles.
+            If None, ``png`` will be used.
+            By default None.
+        show_progress : bool, optional
+            Whether or not to show progress bar, by default False.
+        """
+        self.tile_servers = tile_servers
+        self.temp_folder = "_tile_cache/"
```
[the remark this answers is not in the document]
Comment (rwood-97) [reply mapreader/download/tile_loading.py]: Have added this as a comment in the code next to DEFAULT_TEMP_FOLDER.
In all cases the self.temp_folder is set using DEFAULT_TEMP_FOLDER now so this should be fine if a user/developer wants to change this in the code.

**Review by andrewphilipsmith — APPROVED** [review]
Looks good! Great work

**Pull request merged**